File: tasks-core-jvm2/src/main/java/org/funfix/tasks/UncaughtExceptionHandler.java
package org.funfix.tasks;

import org.jspecify.annotations.NullMarked;

@NullMarked
public final class UncaughtExceptionHandler {
    public static void logException(final Throwable e) {
        final var thread = Thread.currentThread();
        var logger = thread.getUncaughtExceptionHandler();
        if (logger == null) {
            logger = Thread.getDefaultUncaughtExceptionHandler();
        }
        if (logger == null) {
            logger = (t, e1) -> e1.printStackTrace(System.err);
        }
        logger.uncaughtException(thread, e);
    }
}

File: tasks-core/src/jvmMain/kotlin/org/funfix/tasks/UncaughtExceptionHandler.kt
package org.funfix.tasks

import org.jetbrains.annotations.NonBlocking

/**
 * Logs an uncaught exception.
 *
 * On top of the JVM, this function will use the default uncaught exception
 * handler of the current thread. On top of JS, it will use `console.log`.
 */
object UncaughtExceptionHandler {
    @NonBlocking
    @JvmStatic fun rethrowIfFatal(e: Throwable) {
        when (e) {
            is StackOverflowError -> return
            is Error -> throw e
        }
    }

    @NonBlocking
    @JvmStatic fun logOrRethrow(e: Throwable) {
        rethrowIfFatal(e)
        val thread = Thread.currentThread()
        val logger: Thread.UncaughtExceptionHandler =
            thread.uncaughtExceptionHandler
                ?: Thread.getDefaultUncaughtExceptionHandler()
                ?: Thread.UncaughtExceptionHandler { _, it -> it.printStackTrace(System.err) }
        logger.uncaughtException(thread, e)
    }
}
